src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +113,246 @@ object RuleUtils {
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    if (HyperspaceConf.hybridScanEnabled(spark)) {
+      getHybridScanIndexPlan(spark, index, plan, useBucketSpec)
+    } else {
+      getIndexPlan(spark, index, plan, useBucketSpec)
+    }
+  }
+
+  /**
+   * Get alternative logical plan of the current plan using the given index.
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Alternative logical plan.
+   */
+  private def getIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    // Here we can't replace the plan completely with the index. This will create problems.
+    // For e.g. if Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
+    // if we replace this plan with index Scan (A,B,C), we lose the Filter(C=10) and it will
+    // lead to wrong results. So we only replace the base relation.
+    plan transformDown {
+      case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
+        val location =
+          new InMemoryFileIndex(spark, index.content.files, Map(), None)
+        val relation = HadoopFsRelation(
+          location,
+          new StructType(),
+          StructType(index.schema.filter(baseRelation.schema.contains(_))),
+          if (useBucketSpec) Some(index.bucketSpec) else None,
+          new ParquetFileFormat,
+          Map())(spark)
+
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+  }
+
+  /**
+   * Get alternative logical plan of the current plan using the given index.
+   * With HybridScan, indexes with newly appended files to its source relation are also
+   * eligible and we reconstruct new plans for the appended files so as to merge into
+   * the index data.
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Alternative logical plan.
+   */
+  private def getHybridScanIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    // Other than "parquet" format, we cannot read source files along with index data
+    // files in parquet format with 1 FileScan node. In this case, we need to read the
+    // appended source files by another FileScan node and merge into the index data.
+    // Though BucketUnion (using BucketSpec and on-the-fly Shuffle) is used to merge
+    // them for now, we will try to optimize these plans with Union operator.
```

Comment:
```suggestion
    // Since the index data is in "parquet" format, we cannot read source files in 
    // formats other than "parquet" using 1 FileScan node (as the operator requires
    // files in one homogenous format). 
    // 
    // To address this, we need to read the appended source files using another 
    // FileScan node injected into the plan and subsequently merge the data into 
    // the index data.
    //
    // Though BucketUnion (using BucketSpec and on-the-fly Shuffle) is used to merge
    // them for now, these plans will be optimized with a Union operator at a later time
    // (see #145).
```